File: src/main/java/nieboczek/lifestolen/serializer/base/BooleanSerializer.java
package nieboczek.lifestolen.serializer.base;

import nieboczek.lifestolen.serializer.lang.SerializedStringBuilder;
import nieboczek.lifestolen.serializer.lang.TokenStream;
import nieboczek.lifestolen.serializer.lang.TokenType;

public class BooleanSerializer extends Serializer<Boolean> {
    private BooleanSerializer() {
    }

    public static BooleanSerializer of() {
        return new BooleanSerializer();
    }

    @Override
    public void serialize(Boolean value, SerializedStringBuilder builder) {
        builder.text(value.toString());
    }

    @Override
    public Boolean deserialize(TokenStream stream) {
        return Boolean.parseBoolean(stream.nextTokenText(TokenType.IDENTIFIER));
    }
}

File: src/main/kotlin/nieboczek/lifestolen/serializer/base/BooleanSerializer.kt
package nieboczek.lifestolen.serializer.base

import nieboczek.lifestolen.serializer.SerializerError
import nieboczek.lifestolen.serializer.lang.SerializedStringBuilder
import nieboczek.lifestolen.serializer.lang.TokenStream
import nieboczek.lifestolen.serializer.lang.TokenType

class BooleanSerializer : Serializer<Boolean>() {
    override fun serialize(value: Boolean, builder: SerializedStringBuilder) {
        builder.text(value.toString())
    }

    override fun deserialize(stream: TokenStream): Boolean {
        val text = stream.nextTokenText(TokenType.IDENTIFIER)
        val bool = text.toBooleanStrictOrNull()
        return bool ?: throw SerializerError("Invalid boolean value: \"$text\"")
    }
}
